Multibattle testing system

grintoul1 opened the pull request:
Multi-battle Testing System

## Description
Added test types "BATTLE_TEST_MULTI, BATTLE_TEST_AI_MULTI, BATTLE_TEST_TWO_VS_ONE, BATTLE_TEST_AI_TWO_VS_ONE" and definted associated "MULTI_BATTLE_TEST"  and "BATTLE_TEST_ARGS_MULTI" etc

Added BATTLE_TEST_ARGS_MULTI and BATTLE_TEST_ARGS_TWO_VS_ONE - possibly redundant but included in case future development demands their presence
Added MultiBattleTestFunction and TwoVsOneBattleTestFunction, as above perhaps redundant.

Changed IsDoubleBattle() to return true for BATTLE_TYPE_MULTI  and BATTLE_TYPE_INGAME_PARTNER without BATTLE_TYPE_DOUBLE present. All battle initiations set doubles alongside multi and ingame when required, however this now means that extra double flag isn't required.

Changed existing PLAYER( and OPPONENT( defines to use B_POSITION_PLAYER/OPPONENT_LEFT instead of side for consistency - this doesn't impact existing tests
Added MULTI_PLAYER/PARTNER/OPPONENT_A/B defines linked to OpenPokemonMulti() to correspond to appropriate battle position

OpenPokemonMulti assigns mons to appropriate party slot

NOTE: currently this is dependent on player mons being set before partner, and OppA before OppB, else party order messes up. More intelligent solution may be possible

Added series of checks to various DrawTrainerPic functions so the 4 test types produce Brendan, Steven, Leaf, and Red instead of Hiker sprites, and so sprites are positioned appropriately for battle types

Changed InitBtlControllersInternal so these test types are pass the if to be set as Player Primary. Player 2 section changed to InGame & IsRecorded sets controller to RecordedPartner

CURRENT ISSUE: this results in freezes after mon send-in in all tests. When using regular PlayerPartner, partner mons use struggle instead of defined moves.

TESTING check added to TrySetLinkBattleTowerEnemyPartyLevel to block setting of opponent team to L50 as battles run through the Battle Tower Partner Multibattle functions

 

## Issue(s) that this PR fixes
Current testing system only works for Single and Double battle tests. PR aim is to enable 2v2 and 2v1 multi testing

## Feature(s) this PR does NOT handle:
Player vs 2 opponents (yet)

## Discord contact info
grintoul

hedara90 commented on src/battle_script_commands.c:
```diff
@@ -5324,7 +5324,7 @@ static void Cmd_checkteamslost(void)
     if (NoAliveMonsForPlayer())
         gBattleOutcome |= B_OUTCOME_LOST;
-
+    
```
```suggestion

```

hedara90 commented on src/battle_util.c:
```diff
@@ -1571,6 +1571,7 @@ u32 CheckMoveLimitations(u32 battler, u8 unusableMoves, u16 check)
 
     gPotentialItemEffectBattler = battler;
 
+
```
```suggestion
```

hedara90 commented on src/battle_tower.c:
```diff
@@ -3555,18 +3557,21 @@ bool32 ValidateBattleTowerRecord(u8 recordId) // unused
+            s32 i;
+            u8 enemyLevel = SetFacilityPtrsGetLevel();
+
+            for (i = 0; i < PARTY_SIZE; i++)
```
Declare loop iterations in the loop if possible

hedara90 commented on src/battle_message.c:
```diff
@@ -2229,7 +2241,19 @@ void BufferStringBattle(enum StringID stringID, u32 battler)
+                    if (TESTING && gBattleTypeFlags & BATTLE_TYPE_MULTI)
+                    {
+                        if (gBattleScripting.battler == 1)
+                            stringPtr = sText_Trainer1SentOutPkmn;
```
```suggestion
                        }
                            stringPtr = sText_Trainer1SentOutPkmn;
                        }
```
All parts of an if-else chain should have brackets if one of the parts have them.

hedara90 commented on src/battle_message.c:
```diff
@@ -2092,7 +2092,19 @@ void BufferStringBattle(enum StringID stringID, u32 battler)
+                                stringPtr = sText_TwoTrainersWantToBattle;
+                        }
+                        else if(!(gBattleTypeFlags & BATTLE_TYPE_TWO_OPPONENTS))
+                            stringPtr = sText_LinkTrainerWantsToBattlePause;
+                        else
+                            stringPtr = sText_TwoLinkTrainersWantToBattlePause;
```
```suggestion
                        {
                            stringPtr = sText_LinkTrainerWantsToBattlePause;
                        }
                        else
                        {
                            stringPtr = sText_TwoLinkTrainersWantToBattlePause;
                        }
```
All parts of an if-else chain should have brackets if one of the parts have

hedara90 commented on src/battle_controllers.c:
```diff
@@ -2353,7 +2368,7 @@ void BtlController_HandleDrawTrainerPic(u32 battler, u32 trainerPicId, bool32 is
     else
         gSprites[gBattleStruct->trainerSlideSpriteIds[battler]].callback = SpriteCB_TrainerSlideIn;
-
+    
```
```suggestion

```

hedara90 commented on src/battle_controller_recorded_player.c:
```diff
@@ -276,43 +277,49 @@ static void RecordedPlayerHandleDrawTrainerPic(u32 battler)
+                yPos = 80;
+            }
+            else
+                yPos = (8 - gTrainerBacksprites[trainerPicId].coordinates.size) * 4 + 80;
```
```suggestion
            {
                yPos = (8 - gTrainerBacksprites[trainerPicId].coordinates.size) * 4 + 80;
            }
```

hedara90 commented on src/battle_controller_player.c:
```diff
@@ -1897,9 +1904,7 @@ static void PlayerHandleDrawTrainerPic(u32 battler)
         isFrontPic = TRUE;
     }
     else // Use back pic in all the other usual circumstances.
-    {
         isFrontPic = FALSE;
-    }
```
The brackets should be there

hedara90 commented on src/battle_controller_opponent.c:
```diff
@@ -371,18 +373,38 @@ static u32 OpponentGetTrainerPicId(u32 battlerId)
+            else // first mon
+                xPos = 200;
+        }
+        else
+            xPos = 176;
```
```suggestion
        else
        {
            xPos = 176;
        }
```

hedara90 commented on include/test/battle.h:
```diff
@@ -903,6 +955,33 @@ void GigantamaxFactor_(u32 sourceLine, bool32 gigantamaxFactor);
+        else
+            return FALSE;
+    }
+    else
+        return FALSE;
```
```suggestion
    else
    {
        return FALSE;
    }
```

hedara90 reviewed (commented):
Mostly style comments

hedara90 commented:
There are more style things in this.
`if(thing)` should be `if (thing)`.

grintoul1 commented:
_Think_ that's the review comments implemented

grintoul1 commented:
Thought I pushed that last commit this morning sorry!

grintoul1 commented:
Hey @hedara90, do I have any outstanding actions that you need me to do on this at the moment? 🙂

hedara90 commented:
> Hey @hedara90, do I have any outstanding actions that you need me to do on this at the moment? 🙂

Not really, it needs proper reviewing

hedara90 commented:
Please merge current upcoming into this PR, there's been an issue with the test runner that made some tests not run.

hedara90 commented on test/test_runner_battle.c:
```diff
@@ -262,26 +357,97 @@ static void BattleTest_Run(void *data)
+        DATA.recordedBattle.opponentA = TRAINER_LEAF;
         DATA.recordedBattle.opponentB = TRAINER_RED;
         DATA.hasAI = TRUE;
+        DATA.currentMonIndexes[0] = 0;
+        DATA.currentMonIndexes[1] = 0;
+        DATA.currentMonIndexes[2] = 3;
+        DATA.currentMonIndexes[3] = 3;
```
Could you add clarifying comments to these lines (and the other ones where the indices are set manually).
Like:
```
        DATA.currentMonIndexes[0] = 0; // Player first mon
        DATA.currentMonIndexes[1] = 0; // Player partner first mon
        DATA.currentMonIndexes[2] = 3; // Opponent A first mon
        DATA.currentMonIndexes[3] = 3; // Opponent B first mon
```
I might not be too necessary now, but for a future rework that splits multi battle parties it'd be nice to have.

hedara90 commented on test/test_runner_battle.c:
```diff
@@ -2503,6 +2771,7 @@ void SkipTurn(u32 sourceLine, struct BattlePokemon *battler)
 
 void SendOut(u32 sourceLine, struct BattlePokemon *battler, u32 partyIndex)
 {
+    Test_MgbaPrintf("partyIndex %d", partyIndex);
```
Should this be here, or it leftover from testing?

grintoul1 commented on include/battle.h:
```diff
@@ -1222,7 +1222,7 @@ static inline struct PartyState *GetBattlerPartyState(u32 battler)
 static inline bool32 IsDoubleBattle(void)
 {
-    return gBattleTypeFlags & BATTLE_TYPE_DOUBLE;
+    return (gBattleTypeFlags & BATTLE_TYPE_DOUBLE || gBattleTypeFlags & BATTLE_TYPE_MULTI || gBattleTypeFlags & BATTLE_TYPE_INGAME_PARTNER || gBattleTypeFlags & BATTLE_TYPE_TWO_OPPONENTS);
```
Added the below, please let me know if happy or further change needed :)
`#define BATTLE_TYPE_MORE_THAN_TWO_BATTLERS  (BATTLE_TYPE_DOUBLE  || BATTLE_TYPE_MULTI || BATTLE_TYPE_INGAME_PARTNER || BATTLE_TYPE_TWO_OPPONENTS)`

```
static inline bool32 IsDoubleBattle(void)
{
    return (gBattleTypeFlags & BATTLE_TYPE_MORE_THAN_TWO_BATTLERS);
}
```

hedara90 replied:
Yea, it seems fine.

grintoul1 commented on include/test/battle.h:
```diff
@@ -920,6 +1027,38 @@ void GigantamaxFactor_(u32 sourceLine, bool32 gigantamaxFactor);
 void TeraType_(u32 sourceLine, u32 teraType);
 void Shadow_(u32 sourceLine, bool32 isShadow);
 
+static inline bool8 IsMultibattleTest(void)
```
Apologies can I ask for a more detailed explanation?

grintoul1 replied:
Gotcha; question please: doesn't using the `|` when making the composite flag mean it will return true when only one flag is set though? apologies if I'm misunderstanding but I would have thought it'd need some `&&`s in there. Is that not the case?

hedara90 replied:
No, `|` is the bitwise or.
You're turning `0b0001 | 0b1000` into `0b1001` with a single application of it.
Then you doing (for example) `if ((0b1101 & 0b1001) == `0b1001`).

grintoul1 replied:
Removed the `&& TRAINER_BATTLE_PARAM.opponentB == 0xFFFF` as it gave an undeclared error and everything was working fine without it

hedara90 commented on test/test_runner_battle.c:
```diff
@@ -117,7 +191,19 @@ static void InvokeTestFunction(const struct BattleTest *test)
     case BATTLE_TEST_DOUBLES:
     case BATTLE_TEST_AI_DOUBLES:
-        InvokeDoubleTestFunctionWithStack(STATE->results, STATE->runParameter, &gBattleMons[B_POSITION_PLAYER_LEFT], &gBattleMons[B_POSITION_OPPONENT_LEFT], &gBattleMons[B_POSITION_PLAYER_RIGHT], &gBattleMons[B_POSITION_OPPONENT_RIGHT], test->function.singles, &DATA.stack[BATTLE_TEST_STACK_SIZE]);
+        InvokeDoubleTestFunctionWithStack(STATE->results, STATE->runParameter, &gBattleMons[B_POSITION_PLAYER_LEFT], &gBattleMons[B_POSITION_OPPONENT_LEFT], &gBattleMons[B_POSITION_PLAYER_RIGHT], &gBattleMons[B_POSITION_OPPONENT_RIGHT], test->function.doubles, &DATA.stack[BATTLE_TEST_STACK_SIZE]);
```
Something to get MGriffin's eyes on then.

grintoul1 replied:
I've requested an additional review :)

grintoul1 commented on include/test/battle.h:
```diff
@@ -903,6 +955,33 @@ void GigantamaxFactor_(u32 sourceLine, bool32 gigantamaxFactor);
+            return TRUE;
+        else
+            return FALSE;
+    }
```
This section was replaced withthe below following Hedara's review:

```
if (TESTING)
    {
        if (((gBattleTypeFlags & BATTLE_MULTI_TEST) == BATTLE_MULTI_TEST)
        || ((gBattleTypeFlags & BATTLE_TWO_VS_ONE_TEST) == BATTLE_TWO_VS_ONE_TEST))
            return TRUE;
        else
            return FALSE;
    }
```

grintoul1 commented:
Summary of last changes:
1. Added `AI_PARTNER_NAME` to `include/test/battle.h` for `SCENE` use
2. Added `TestRunner_Battle_CheckSwitch` to `PlayerPartnerHandleChoosePokemon` if `TESTING`
3. `InitBtlControllersInternal` changed to only set RecordedPartner for non-AI tests, else use PlayerPartner
4. Strings added to `battle_message.c` for sending out and withdrawing mons so partner trainer name used. Can put these behind an `if (TESTING)` if desired, but it does seem odd that partner withdraws and send outs use the player-style text.
5. Moved AI multi tests (non-switching) into a new `ai_multi.c` file
6. Handful of tests added to `ai_switching.c` to test `EXPECT_SEND_OUT` and `EXPECT_SWITCH` for multi partner
7. Amended `ExpectSendOut` invalid text to refer to new test types and to not force Celebrate if no input given for battler2 in AI multi tests

hedara90 commented on src/battle_controller_player_partner.c:
```diff
@@ -310,6 +318,9 @@ static void PlayerPartnerHandleChoosePokemon(u32 battler)
     }
+    #if TESTING
+    TestRunner_Battle_CheckSwitch(battler, chosenMonId);
+    #endif // TESTING
```
Not right now, the non-preproc thing started after the testing system was created.
Changing it should probably happen in a specific cleanup PR.
